modules/renter.go

```diff
@@ -50,12 +54,92 @@ type Allowance struct {
 // DownloadInfo provides information about a file that has been requested for
 // download.
 type DownloadInfo struct {
-	SiaPath     string    `json:"siapath"`
-	Destination string    `json:"destination"`
-	Filesize    uint64    `json:"filesize"`
-	Received    uint64    `json:"received"`
-	StartTime   time.Time `json:"starttime"`
-	Error       string    `json:"error"`
+	SiaPath     string         `json:"siapath"`
+	Destination DownloadWriter `json:"destination"`
+	Filesize    uint64         `json:"filesize"`
+	Received    uint64         `json:"received"`
+	StartTime   time.Time      `json:"starttime"`
+	Error       string         `json:"error"`
+}
+
+// DownloadWriter provides an interface which all output writers have to implement.
+type DownloadWriter interface {
+	WriteAt(b []byte, off int64) (int, error)
+}
+
+// DownloadFileWriter is a file-backed implementation of DownloadWriter.
+type DownloadFileWriter struct {
+	f        *os.File
+	Location string
+	offset   uint64
+}
+
+// NewDownloadFileWriter creates a new instance of a DownloadWriter backed by the file named.
+func NewDownloadFileWriter(fname string, offset, length uint64) *DownloadFileWriter {
+	l, _ := os.OpenFile(fname, os.O_CREATE|os.O_WRONLY, defaultFilePerm)
```

Comment:
What happens if you call `WriteAt()` out of order with the DownloadFileWriter? Going to disk makes sense, but if you are doing it in a stream, it's going to need to block so that `siad` doesn't consume more memory than it is allowed to (if it returns, the calling thread will send off another flight of downloads, consuming an extra 40MB of RAM).

But if it's blocking, it means that you have to make sure that all previous chunks are at the very least in-progress, which puts an unexpected constraint on the download algorithm. It's a constraint that we already fulfill, but one that we'll at least need to be explicit about in the future.

But, there's also some optimization based stuff that we might lose. One thing we are going to do to speed up downloads is filter hosts into fast and slow groups. If you are doing an in-order download, the slow group could bottleneck all the fast groups, because the WriteAt calls will not return until the slow groups complete. This could significantly impact overall performance, which means if you are doing sequential downloads, you may want to disable the slowest groups and just use the fast hosts (depending on how much RAM the user is allowing you to use. If you've got more than 2GB to work with it's probably not going to matter).